test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java

```diff
@@ -0,0 +1,468 @@
+    }
+
+
+
+    private CompletableFuture<Void> startWriting(final EventStreamWriter<Long> writer) {
+        return CompletableFuture.runAsync(() -> {
+            while (!testState.stopWriteFlag.get()) {
+                try {
+                    long value = testState.eventData.incrementAndGet();
+                    Thread.sleep(100);
+                    log.debug("Event write count before write call {}", value);
+                    writer.writeEvent(String.valueOf(value), value);
+                    log.debug("Event write count before flush {}", value);
+                    writer.flush();
+                    testState.eventWriteCount.getAndIncrement();
+                    log.debug("Writing event {}", value);
+                } catch (InterruptedException e ) {
+                        log.error("Error in sleep: ", e);
+                        break;
+                } catch (ConnectionClosedException e) {
+                        log.warn("Test exception in writing events: ", e);
+                        continue;
+                } catch (Throwable e) {
+                    log.error("Test exception in writing events: ", e);
+                        break;
+                }
+            }
+        }, executorService);
+    }
+
+    private CompletableFuture<Void> startReading(final EventStreamReader<Long> reader) {
+        return CompletableFuture.runAsync(() -> {
+            log.info("Exit flag status {}", testState.stopReadFlag.get());
+            log.info("Read count {} and write count {}", testState.eventReadCount.get(), testState.eventWriteCount.get());
+            while (!(testState.stopReadFlag.get() && testState.eventReadCount.get() == testState.eventWriteCount.get())) {
+                log.info("Entering read loop");
+                // exit only if exitFlag is true  and read Count equals write count.
+                try {
+                    final Long longEvent = reader.readNextEvent(SECONDS.toMillis(60)).getEvent();
+                    log.debug("Reading event {}", longEvent);
+                    if (longEvent != null) {
+                        //update if event read is not null.
+                        testState.eventsReadFromPravega.add(longEvent);
+                        testState.eventReadCount.incrementAndGet();
+                        log.debug("Event read count {}", testState.eventReadCount.get());
+                    } else {
+                        log.debug("Read timeout");
+                    }
+                } catch (ConnectionClosedException e ) {
+                    log.warn("Test exception in reading events: ", e);
+                    continue;
+                } catch (ReinitializationRequiredException e) {
+                    log.error("Test exception in reading events: ", e);
+                    Assert.fail("Unable to read events. Test failure");
```

Comment:
Because this is called inside of a runAync block (meaning it will be run from another thread) this call will not actually fail the test. You need to call fail from the main test thread.